File: Cascaded/BEP/FIG2a/mainBEP.m
clear all
close all
clc

% Piece of code to calculate the BEP - Bit Error Probability
% Pedro Henrique Dornelas Almeida - 08/11/2023

functions_path = "functions";
addpath(functions_path);

set(0,'defaulttextinterpreter','latex');
set(groot,'defaultAxesTickLabelInterpreter','latex');

% SNRs -- Amostragem dos valores observáveis
L = 0;    %db
U = 50;   %db
points = 1e2;
bounds = [L U]; %dB gammaBar limits

analit_gammaBar_dB = linspace(L, U, points);
analit_gammaBar = 10.^(0.1*analit_gammaBar_dB);
simu_gammaBar_dB = linspace(L, U, 15);      % SNR em dB
simu_gammaBar = 10.^(0.1*simu_gammaBar_dB); % SNR linear

% Parâmetro da modulação (teóricas)
rho = 1;
% Parâmetro da modulação (simulação)
% Ordem da constelação MQAM
M = 2;

% Parâmetros da Distribuição Alfa F - Cascaded
N = 2;    % nº estações relay
alfa = [2.2, 2.5];
mu = 1;
ms = [2];

analit_gammaBar_c = db2pow(1) * (ones( length(analit_gammaBar) , max(N)));
% analit_gammaBar_c = (ones( length(analit_gammaBar) , max(N)));
analit_gammaBar_c(:, 1) = analit_gammaBar; % variar só do primeiro canal...
simu_gammaBar_c = db2pow(1) * (ones( length(simu_gammaBar) , max(N)));
% simu_gammaBar_c = (ones( length(simu_gammaBar) , max(N)));
simu_gammaBar_c(:, 1) = simu_gammaBar; % variar só do primeiro canal...

% Erro de apontamento
z = [0.6, 0.8; 
     1.0, 1.1; 
     1.5, 1.6; 
     8 ,  9;];
% z = [ 8,  9;
%      14, 15;];
% z = [0.6, 0.8; 1.0, 1.1; 3, 3.3; 14, 15];

for i=1:length(alfa)
    if ms <= (4/alfa(i))
        fprintf('ms > 4/alpha not met\n')
        error
    end
end

% Parâmetros da simulação
Nc = 1e6; % Número de pontos
Ao = 0.8;
% Ao = 1;
rc = 1;

% Perda de percurso
Hl = 1.00;

% início da simulação e geração de gráficos
colorz = 'brgmp';

tic;


figure(1)
cont = 1;


for k = 1:length(z)
    k
    for w = 1:length(ms)
        % simulation_params = [alfa(1), mu, ms(w), z(k, 1), rc, Hl;
        %                      alfa(2), mu, ms(w), z(k, 2), rc, Hl;];

        simulation_params = [alfa(1), mu, ms(w), z(k, 1), Ao, Hl;
                             alfa(2), mu, ms(w), z(k, 2), Ao, Hl;];

        analit_params = [alfa(1), mu, ms(w), z(k, 1);
                         alfa(2), mu, ms(w), z(k, 2);];

        BEP_sim = channel(M, N, simulation_params, Nc, simu_gammaBar_c);        
        [BEP] = BEP_analit(N, analit_params, rho, analit_gammaBar_c);
        [BEP_asy] = BEP_asymptotic(N, analit_params, rho, analit_gammaBar_c);

        h(cont)   = semilogy(analit_gammaBar_dB, BEP, colorz(k),'linewidth',1.2); hold on;
        h(cont+1) = semilogy(analit_gammaBar_dB, BEP_asy,'k--', 'linewidth',1.2);hold on;
        h(cont+2) = semilogy(simu_gammaBar_dB, BEP_sim, [colorz(k),'x'], 'linewidth', 1.2);hold on;
        cont = cont+3;
        
    end
end

execution_time = toc;
disp(['Execution time: ' num2str(execution_time) 's']);

axis([min(simu_gammaBar_dB) max(simu_gammaBar_dB) 1e-5 1])
tam_fonte = 11;
legend('FontSize', tam_fonte, 'Location', 'southwest')
% legend("$z_1="+num2str(z(1,1))+", z_2="+num2str(z(1,2))+"$", '', "$z_1="+num2str(z(2,1))+", z_2="+num2str(z(2,2))+"$", '', "$z_1="+num2str(z(3,1))+", z_2="+num2str(z(3,2))+"$", '', "$z_1="+num2str(z(4,1))+", z_2="+num2str(z(4,2))+"$", 'Asymptotic', 'Location', 'southwest')
% legend("$z_1="+num2str(z(1,1))+", z_2="+num2str(z(1,2))+"$", '', "$z_1="+num2str(z(2,1))+", z_2="+num2str(z(2,2))+"$", '', "$z_1="+num2str(z(3,1))+", z_2="+num2str(z(3,2))+"$", '', "Non-pointing errors", 'Asymptotic', 'Location', 'southwest')
legend([h(1), h(4), h(7), h(10), h(11), h(12)], {"$z_1="+num2str(z(1,1))+", z_2="+num2str(z(1,2))+"$" , "$z_1="+num2str(z(2,1))+", z_2="+num2str(z(2,2))+"$",...
                                                 "$z_1="+num2str(z(3,1))+", z_2="+num2str(z(3,2))+"$" , "Non-pointing errors", "Asymptotic", "Simulated"});
ax = gca;
ax.FontSize = 13;

%title('Outage Probability')
set(legend, 'Interpreter', 'latex')
ylabel("BEP", 'FontSize', 14)
yticks([1e-5, 1e-4, 1e-3, 1e-2, 1e-1, 1e0])
xlabel("SNR (dB)", 'FontSize', 14)
grid on

if rho == 1
    modulation = '(BPSK)';
elseif rho == 0.5
    modulation = '(QPSK)';
end

%textbox com valores
dim = [0.15 0.45 0.2 0.2];
str = {"$\alpha_1 ="+num2str(alfa(1))+"$", "$\alpha_2 ="+num2str(alfa(2))+"$", "$\mu ="+num2str(mu)+"$", "$m ="+num2str(ms)+"$","$\rho="+num2str(rho)+"$ "+modulation};
annotation('textbox',dim,'interpreter','latex','String',str,'FitBoxToText','on', 'FontSize', tam_fonte);

%simu_gammaBar_dB,cdfG(gth(1),gammaBar),'-',

File: RIS/BEP_new/main_BEP.py
import matplotlib.pyplot as plt
import numpy as np
from scipy.io import savemat, loadmat

from BEP_analit import BEP_analit
from BEP_asymptotic import BEP_asymptotic

fig, ax = plt.subplots()

L = [1,2,3,4]
N = 2
alpha = [1.5, 2.3]
ms = [4, 5]
mu = [2.5, 3.5]
# z = [[0.7, 0.72],
#      [  7,   8]]

z = [[0.85, 0.87],
     [  7,   8]]

p = 1    # BPSK modulation

for idx, a in enumerate(alpha):
    if ms[idx] <= (4/a):
        print("ms > 4/alpha not met. Exiting...")
        exit()

points = 100
l_bound_dB = 0
u_bound_dB = 30
gamma_bar_dB = np.linspace(l_bound_dB, u_bound_dB, points)
gamma_bar = 10 ** (gamma_bar_dB / 10)

analit_gamma_bar_c = (10 ** (1 / 10)) * np.ones((points, N))
analit_gamma_bar_c[:,0] = gamma_bar
# print(analit_gamma_bar_c)

BEP = np.zeros((points, len(L), len(z)))
BEP_asy = np.zeros((points, len(L), len(z)))

color = ['b','g','r','m']

for i in range(len(L)):
    for k in range(len(z)):
        analit_params = [[alpha[0], mu[0], ms[0], z[k][0]],
                         [alpha[1], mu[1], ms[1], z[k][1]]]
        # print(analit_params)

        BEP[:, i, k] = BEP_analit(L[i], N, analit_params, p, analit_gamma_bar_c)
        # print(BEP[:, i, k])
        BEP_asy[:, i, k] = BEP_asymptotic(L[i], N, analit_params, p, analit_gamma_bar_c)

        if k == 0:
            ax.semilogy(gamma_bar_dB, BEP[:, i, k], color=color[i], linestyle='-', label=(f'L = {L[i]}'))
        else:
            ax.semilogy(gamma_bar_dB, BEP[:, i, k], color=color[i], linestyle='-')
        
        if i == len(L)-1 and k == len(z)-1:
            ax.semilogy(gamma_bar_dB, BEP_asy[:, i, k], color='k', linestyle='--', label='Asymptotic')
        else:
            ax.semilogy(gamma_bar_dB, BEP_asy[:, i, k], color='k', linestyle='--')

    print(L[i])

filename = "BEP.mat"
savemat(filename, dict(L=L,
                       N=N,
                       gamma_bar_dB=gamma_bar_dB,
                       BEP=BEP,
                       BEP_asy=BEP_asy,
                       alpha=alpha,
                       mu=mu,
                       ms=ms,
                       z=z,
                       p=p))
print("points saved...")


ax.legend(loc='lower left')
ax.set_ylabel("BEP")
ax.set_xlabel("SNR (dB)")
ax.grid(linestyle='--', axis='both', linewidth=0.5)
plt.ylim([10**(-5), 10**0])
plt.xlim([min(gamma_bar_dB), max(gamma_bar_dB)])
plt.show()